File: PDOO/Practica3/Ruby/Irrgarten/game.rb
require_relative 'dice'
require_relative 'player'
require_relative 'monster'
require_relative 'labyrinth'
require_relative 'game_state'
require_relative 'game_character'
require_relative 'orientation'

module Irrgarten


    class Game
        @@MAX_ROUNDS = 10

        #------------------------|Custom Labyrinth|------------------------
        @@ROWS = 15
        @@COLS = 15

        @@NMONSTERS = 5
        @@MONSTERS_POS = [[0, 0], [1, 6], [2, 2], [5, 5], [7, 1]]

        @@NBLOCKS = 4
        @@BLOCKS_POS = [
            [Orientation::HORIZONTAL, 0, 1, 4],
            [Orientation::HORIZONTAL, 9, 0, 2],
            [Orientation::VERTICAL, 2, 8, 7],
            [Orientation::VERTICAL, 3, 3, 4]
        ]
        #------------------------------------------------------------------

        def initialize(n_players)
            exit_row = Dice.random_pos(@@ROWS)
            exit_col = Dice.random_pos(@@COLS)

            @players = Array.new
            @monsters = Array.new

            n_players.times do |i|
                @players.push(Player.new(i, Dice.random_intelligence, Dice.random_strength))
            end

            @current_player_index = Dice.who_starts(n_players)
            @current_player = @players[@current_player_index]

            @labyrinth = Labyrinth.new(@@ROWS, @@COLS, exit_row, exit_col)
            configure_labyrinth
            @labyrinth.spread_players(@players)

            @log = "\n╔════════════════════════════╗\n" +
                   "║    The Game has started!   ║\n" +
                   "╚════════════════════════════╝\n\n";
        end

        def next_step(preferred_direction)
            dead = @current_player.dead
            direction = actual_direction(preferred_direction)

            if (!dead)
                if (direction != preferred_direction)
                    self.log_player_no_orders
                end

                monster = (@labyrinth.put_player(direction, @current_player))
                if (monster == nil)
                    self.log_no_monster
                else
                    winner = self.combat(monster)
                    self.manage_reward(winner)
                end
            else
                self.manage_resurrection
            end

            end_game = self.finished
            if (!end_game)
                self.next_player
            end
            return end_game
        end


        def finished
            @labyrinth.have_a_winner
        end

        def game_state
            players_str = ""
            @players.each do |player|
                players_str += player.to_s + "\n"
            end

            monsters_str = ""
            @monsters.each do |monster|
                monsters_str += monster.to_s + "\n"
            end

            GameState.new(@labyrinth.to_s, players_str, monsters_str, @current_player_index, self.finished, @log)
        end

        def configure_labyrinth

            @@BLOCKS_POS.each do |block|
                orientation, row, col, size = block
                @labyrinth.add_block(orientation, row, col, size)
            end

            @@MONSTERS_POS.each do |pos|
                row, col = pos
                @monsters.push(Monster.new("Monster " + pos.to_s, Dice.random_intelligence, Dice.random_strength))
                @labyrinth.add_monster(row, col, @monsters.last)
            end
        end

        def next_player
            @current_player_index = (@current_player + 1) % players.size
            @current_player = @players[@current_player_index]
        end

        def actual_direction(preferred_direction)
            current_row = @current_player.row
            current_col = @current_player.col

            valid_moves = labyrinth.valid_moves(current_row,current_col)
            current_player.move(preferred_direction, valid_moves)
        end

        def combat(monster)
            rounds = 0
            winner = GameCharacter::PLAYER
            lose = monster.defend(@current_player.attack)

            while (!lose && rounds < @@MAX_ROUNDS) do
                winner = GameCharacter::MONSTER
                rounds += 1

                lose = @current_player.defend(monster.attack)

                if (!lose)
                    winner = GameCharacter::PLAYER
                    lose = monster.defend(@current_player.attack)
                end
            end

            winner
        end

        def manage_reward(winner)
            if (winner == GameCharacter::PLAYER)
                @current_player.receive_reward
                log_player_won
            else
                log_monster_won
            end
        end

        def manage_resurrection
            if (Dice.resurrect_player)
                @current_player.resurrect
                log_resurrected
            else
                log_player_skip_turn
            end
        end

        def log_player_won
            log += "Player #{@current_player_index} has won.\n"
        end

        def log_monster_won
            log += "A Monster has won the fight.\n"
        end

        def log_resurrected
            log += "Player #{@current_player_index} has resurrected.\n"
        end

        def log_player_skip_turn
            log += "Player #{@current_player_index} has skipped the turn.\n"
        end

        def log_player_no_orders
            log += "Player #{@current_player_index} didn't follow the order.\n"
        end

        def log_no_monster
            log += "Player #{@current_player_index} has moved to an empty cell or it wasn't possible to move.\n"
        end

        def log_rounds(rounds, max)
            log += "Rounds: #{rounds}|#{max}.\n"
        end
    end
end

File: PDOO/Practica4/Java/Irrgarten/src/irrgarten/Game.java
/**
 * @file Game.java
 * @author Miguel Angel De la Vega Rodríguez
 * @description This file contains the Game class, which provides methods
 * to manage the game, such as moving the players, managing the turns, and
 * checking if the game has finished
 */

package irrgarten;

import java.util.ArrayList;

public class Game {
    private static final int MAX_ROUNDS = 10;
    
    /*----------------------------Custom Labyrinth----------------------------*/
    private static final int ROWS = 11;
    private static final int COLS = 10;
    
    private static final int NMONSTERS = 5;
    private static final int[][] MONSTERS_POS = {{0,0},{1,6},{2,2},{5,5},{7,1}};
    
    private static final int NBLOCKS = 4;
    private static final Object[][] BLOCKS_POS = {{Orientation.HORIZONTAL,0,1,4},
                                                  {Orientation.HORIZONTAL,9,0,2},
                                                  {Orientation.VERTICAL,2,8,7},
                                                  {Orientation.VERTICAL,3,3,4}};
    
    private int currentPlayerIndex;
    private Player currentPlayer;
    private String log;
    private Labyrinth labyrinth;
    private ArrayList<Player> players;
    private ArrayList<Monster> monsters;

    /**
     * Constructor for the game, it receives the number of players and sets up
     * the labyrinth for the players to play 
     * 
     * @param nPlayers number of players
     */
    public Game(int nPlayers) {
        int exitRow = Dice.randomPos(ROWS);
        int exitCol = Dice.randomPos(COLS);
        
        
        players = new ArrayList<>();
        monsters = new ArrayList<>();
        
        for (int playerNum = 0; playerNum < nPlayers; ++playerNum) {
            players.add(new Player(Character.forDigit(playerNum,10), Dice.randomIntelligence(), Dice.randomStrength()));
        }
        
        currentPlayerIndex = Dice.whoStarts(nPlayers);
        currentPlayer = players.get(currentPlayerIndex);
        
        labyrinth = new Labyrinth(ROWS, COLS, exitRow, exitCol);
        configureLabyrinth();
        labyrinth.spreadPlayers(players);
        
        
        log = "╔════════════════════════════╗\n" +
              "║    The Game has started!   ║\n" +
              "╚════════════════════════════╝\n\n";
        
    }

    /**
     * Returns whether the game has finished or not using the haveAWinner
     * method of the labyrinth
     *
     * @return true if the game has finished, false otherwise
     */
    public boolean finished() {
        return this.labyrinth.haveAWinner();
    }

    /**
     * It manages a whole turn in the game.
     * 
     * @param preferredDirection direction in which the player prefers to go
     * @return Whether the game ended or not
     */
    public boolean nextStep(Directions preferredDirection) {
        boolean dead = currentPlayer.dead();
        log = "";

        if (!dead) {
            Directions direction = actualDirection(preferredDirection);
            if (direction != preferredDirection) {
                logPlayerNoOrders();
            }
            Monster monster = labyrinth.putPlayer(direction, currentPlayer);
            
            if (monster == null) {
                logNoMonster();
            }
            else{
                GameCharacter winner = combat(monster);
                manageReward(winner);
            }

        }
        else{
            manageResurrection();
        }

        boolean endGame = finished();

        if (!endGame) {
            nextPlayer();
        }
        
        return endGame;
    }

    /**
     * Returns the current state of the game as a GameState object formatted
     * 
     * @return current state of the game
     */
    public GameState getGameState() {
        String player="";
        String monster= "";
        
        for (int i = 0; i < players.size(); ++i) {
            player += players.get(i).toString() + "\n";
        }
        for (int i = 0; i < monsters.size(); ++i) {
            monster += monsters.get(i).toString()+ "\n";
        }

        return new GameState(this.labyrinth.toString(),player,monster,
                this.currentPlayerIndex,this.finished(),this.log);
    }

    /**
     * Configures the labyrinth for the game, it adds obstacles and monsters 
     * saving them in this class
     *
     */
    private void configureLabyrinth() {
        for (int i = 0; i < NMONSTERS; ++i) {
            Monster monster = new Monster("Monster " + i, Dice.randomIntelligence(),
                Dice.randomStrength());
            
            monsters.add(monster);
            labyrinth.addMonster(MONSTERS_POS[i][0], MONSTERS_POS[i][1], monster);
        }
        
        for (int i = 0; i < NBLOCKS; ++i) {
            labyrinth.addBlock((Orientation)BLOCKS_POS[i][0], (int)BLOCKS_POS[i][1],
                    (int)BLOCKS_POS[i][2], (int)BLOCKS_POS[i][3]);
        }
    }

    /**
     * Updates both the attributes that store the current player and the index 
     * of the current player
     */
    private void nextPlayer() {
        currentPlayerIndex = (currentPlayerIndex + 1) % players.size();
        currentPlayer = players.get(currentPlayerIndex);
    }

    /**
     * Returns the actual direction where the player moves considering the
     * preferred direction 
     * 
     * @param preferredDirection direction which the player prefers to go to
     * @return Actual direction where the players goes
     */
    private Directions actualDirection(Directions preferredDirection) {
        int currentRow = currentPlayer.getRow();
        int currentCol = currentPlayer.getCol();

        ArrayList<Directions> validMoves = labyrinth.validMoves(currentRow, currentCol);

        return currentPlayer.move(preferredDirection, validMoves);
    }

    /**
     * This method is responsible for managing combat between a player
     * and a monster, it returns the winner of the combat
     * 
     * @param monster monster which is in combat with current player
     * @return who wins
     */
    private GameCharacter combat(Monster monster) {
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        boolean lose = monster.defend(currentPlayer.attack());

        while (!lose && rounds < MAX_ROUNDS) {
            winner = GameCharacter.MONSTER;
            rounds++;

            lose = currentPlayer.defend(monster.attack());

            if (!lose) {
                winner = GameCharacter.PLAYER;
                lose = monster.defend(currentPlayer.attack());
            }
        }
        
        logRounds(rounds,MAX_ROUNDS);
        return winner;
    }

    /**
     * This method manages the rewards produced when finished a combat
     * it also logs whether one of them has won
     * 
     * @param winner of the combat
     */
    private void manageReward(GameCharacter winner) {
        if (winner == GameCharacter.PLAYER) {
            currentPlayer.receiveReward();
            logPlayerWon();
        }
        else{
            logMonsterWon();
        }
    }

    /**
     * This method manages the resurrection of a player
     */
    private void manageResurrection() {
        if (Dice.resurrectPlayer()) {
            currentPlayer.resurrect();
            logResurrected();
        }
        else{
            logPlayerSkipTurn();
        }
    }

    /**
     * Appends to the end of the attribute log the message indicating
     * that the player has won, it also adds a new line at the end
     */
    private void logPlayerWon() {
        log += "The player has won\n";
    }

    /**
     * Appends to the end of the attribute log the message indicating 
     * that the monster has won, it also adds a new line at the end
     */
    private void logMonsterWon() {
        log += "The monster has won\n";
    }

    /**
     * Appends to the end of the attribute log the message indicating
     * that the player has been resurrected, it also adds a new line at the end
     */
    private void logResurrected() {
        log += "The player has been resurrected\n";
    }

    /**
     * Appends to the end of the attribute log the message indicating
     * that the player has skipped his turn for being dead, it also adds a new line at the end
     */
    private void logPlayerSkipTurn() {
        log += "The player has skipped his turn for being dead\n";
    }

    /**
     * Appends to the end of the attribute log the message indicating
     * that the player has not followed the orders as it was not possible
     */
    private void logPlayerNoOrders() {
        log += "The player hasn't followed the orders\n";
    }

    /**
     * Appends to the end of the attribute log the message indicating
     * that the player has moved to an empty cell or has not moved
     * it also adds a new line at the end
     */
    private void logNoMonster() {
        log += "Player has moved to an empty cell or has not moved\n";
    }

    /**
     * Appends to the end of the attribute log the message indicating
     * the rounds 
     */
    private void logRounds(int rounds, int max) {
        log += "Rounds: " + rounds + "|" + max + "\n";
    }
}
